# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webscheduling/WebTaskSchedulerWorker.h
# Commit: 4ccb6970887e
# Full Hash: 4ccb6970887ece212a1176681a567480ea586dfc
# Author: Sean Feng <sefeng@mozilla.com>
# Date: 2025-02-25 09:30:56
# Regressor Bug: 1920115
# File Overlap Count: 2
# Description:
#   Bug 1920115 - Allow high priority tasks to run before timers r=smaug
#   
#   In Gecko, unlike other runnables where the TaskController
#   picks one each at a time, the TimeoutManager allows multiple
#   timers to run sequentially without allowing other runnables
# ==============================================================================

diff -r 7522970c5a1d -r 4ccb6970887e dom/webscheduling/WebTaskSchedulerWorker.h
--- a/dom/webscheduling/WebTaskSchedulerWorker.h	Mon Feb 24 19:31:28 2025 +0000
+++ b/dom/webscheduling/WebTaskSchedulerWorker.h	Mon Feb 24 19:31:28 2025 +0000
@@ -38,6 +38,13 @@
 
   void Disconnect() override;
 
+  void IncreaseNumNormalOrHighPriorityQueuesHaveTaskScheduled() override;
+  void DecreaseNumNormalOrHighPriorityQueuesHaveTaskScheduled() override;
+
+  bool HasScheduledNormalOrHighPriorityWebTasks() const {
+    return mNumHighPriorityQueuesHaveTaskScheduled;
+  }
+
  private:
   ~WebTaskSchedulerWorker() = default;
 
@@ -47,6 +54,12 @@
 
   RefPtr<StrongWorkerRef> mWorkerRef;
   bool mWorkerIsShuttingDown{false};
+
+  // Unlike window global where multiple globals can share the
+  // same event loop, worker globals don't share event loops,
+  // so it's okay to have this counter lives inside the
+  // scheduler for workers.
+  uint32_t mNumHighPriorityQueuesHaveTaskScheduled = 0;
 };
 }  // namespace mozilla::dom
 #endif